mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+- Until the "end access" fence is signaled:
+  - The `GPUDevice` has exclusive, read/write access to the imported buffer
+  - All WebNN work involving the imported `MLTensor` is blocked
+- When the `GPUBuffer` is destroyed, the "end access" fence is signaled and the `MLTensor` may be used again by WebNN
+
+### `compute()` vs. `dispatch()`
+
+`compute()` will be deprecated and removed in favor of `dispatch()`.
+
+It's possible `compute()` may have a performance advantage on some platforms for some use cases, such as in one-shot inference (where the benefits of buffer re-use are not relevant) with small inputs/outputs on CPU (where the overhead of task queueing may outweigh the benefits of parallelization). At this time, we do not expect this performance impact to be substantial enough to justify providing two mostly-identical methods for executing an `MLGraph`.
+
+### Open Questions
+
+- How will errors be surfaced? Do we need a concept similar to [WebGPU's error scopes](https://www.w3.org/TR/webgpu/#error-scopes), or is [returning errors via a promise for select operations sufficient](https://github.com/webmachinelearning/webnn/issues/697#issuecomment-2195656878)? See [#477](https://github.com/webmachinelearning/webnn/issues/477)
+- On non-UMA systems, does the user agent have enough information to appropriately allocate an `MLTensor` if an `MLDeviceType` is not used for creating an `MLContext`? See [#350](https://github.com/webmachinelearning/webnn/issues/350) and [#749](https://github.com/webmachinelearning/webnn/issues/749)
+- Should the `dispatch()` method be a part of the `MLGraph` interface rather than `MLContext`? Should `readBuffer()` and `writeBuffer()` exist on an `MLTensor`? See [#697](https://github.com/webmachinelearning/webnn/issues/697).
+- If an `MLContext` is not created from a `GPUDevice`, does there need to be some mechanism - above and beyond the `MLTensorUsage.WEBGPU_INTEROP` flag - for identifying the specific `GPUDevice` with which interop is desired?
+- What are the usage flags of a `GPUBuffer` created from an `MLTensor`?
+
+## Considered Alternatives
+
+### `MLTensor` as a generic bag of bytes (i.e. `MLBuffer`)
+
+An `MLTensor` was at one point proposed to be a opaque bag of bytes which could be passed as an input or output buffer to a machine learning graph. However, graph inputs and outputs are expressed not as _buffers_ but _tensors_ with a data type and shape. Generic reinterpretation of an opaque buffer is [not a viable approach on platforms which require typed tensors, such as Core ML](https://github.com/webmachinelearning/webnn/issues/542#issuecomment-2067555410). Since the [use case](https://w3ctag.github.io/design-principles/#usecase-oriented-apis) of an `MLTensor` is as a tensor, an `MLTensor` is effectively typed as a tensor.
```

Comment:
@reillyeon and I were discussing potentially requiring `writeBuffer()` (and the in-place variant of `readBuffer()`) to be called with either:
- a `TypedArray` corresponding to `dataType` (e.g. `"int32"` => `Int32Array`), or
- a `Uint8Array` for ease of use with WebAssembly

This seems like a good balance between preventing accidental misuse of the API without requiring WASM callers to implement a bunch of boilerplate code

> especially given JS Float16Array is not mature yet

Good callout. In this model, we would allow `"float16"` as `UInt16Array` (at least for now?)

Reply:
Yeah, we need `UInt16Array` for now for float16 because that's what the samples and [ORT use for input/output tensors](https://github.com/microsoft/webnn-developer-preview/blob/fc15fc9974b3153d311d911ec24da5645687fe64/demos/stable-diffusion-1.5/utils.js#L88).

Reply:
Ack. I've added this to the list of issues I will follow up on once this PR merges....